sklearn/preprocessing/_encoders.py

```diff
@@ -903,13 +911,25 @@ def fit(self, X, y=None):
                     self._missing_indices[cat_idx] = i
                     continue
 
-        if np.dtype(self.dtype).kind != "f" and self._missing_indices:
-            raise ValueError(
-                "There are missing values in features "
-                f"{list(self._missing_indices)}. For OrdinalEncoder to "
-                "passthrough missing values, the dtype parameter must be a "
-                "float"
-            )
+        if self._missing_indices:
+            if np.dtype(self.dtype).kind != "f" and is_scalar_nan(
+                self.encoded_missing_value
+            ):
+                raise ValueError(
+                    "There are missing values in features "
+                    f"{list(self._missing_indices)}. For OrdinalEncoder to "
+                    f"encode missing values with dtype: {self.dtype}, set "
+                    "encoded_missing_value to a non-nan value, or "
+                    "set dtype to a float"
+                )
+
+            if not is_scalar_nan(self.encoded_missing_value):
+                for feature_cats in self.categories_:
+                    if 0 <= self.encoded_missing_value < len(feature_cats):
+                        raise ValueError(
+                            f"encoded_missing_value ({self.encoded_missing_value}) "
+                            "is already used to encode a known category"
+                        )
```

Comment:
If we have `feature_names_in_` defined, this error message should include it. Otherwise adding the feature index would also be helpful.